src/librustdoc/html/format.rs

```diff
@@ -1089,52 +1137,55 @@ impl clean::Visibility {
-            clean::Public => f.write_str("pub "),
-            clean::Inherited => Ok(()),
-
+        let to_print = match self {
+            clean::Public => "pub ".to_owned(),
+            clean::Inherited => String::new(),
             clean::Visibility::Restricted(vis_did) => {
                 // FIXME(camelid): This may not work correctly if `item_did` is a module.
                 //                 However, rustdoc currently never displays a module's
                 //                 visibility, so it shouldn't matter.
                 let parent_module = find_nearest_parent_module(tcx, item_did);
 
                 if vis_did.index == CRATE_DEF_INDEX {
-                    write!(f, "pub(crate) ")
+                    "pub(crate) ".to_owned()
                 } else if parent_module == Some(vis_did) {
                     // `pub(in foo)` where `foo` is the parent module
                     // is the same as no visibility modifier
-                    Ok(())
+                    String::new()
                 } else if parent_module
                     .map(|parent| find_nearest_parent_module(tcx, parent))
                     .flatten()
                     == Some(vis_did)
                 {
-                    write!(f, "pub(super) ")
+                    "pub(super) ".to_owned()
                 } else {
-                    f.write_str("pub(")?;
                     let path = tcx.def_path(vis_did);
                     debug!("path={:?}", path);
                     let first_name =
                         path.data[0].data.get_opt_name().expect("modules are always named");
+                    // modified from `resolved_path()` to work with `DefPathData`
+                    let last_name = path.data.last().unwrap().data.get_opt_name().unwrap();
+                    let anchor = anchor(vis_did, &last_name.as_str(), cache).to_string();
+
+                    let mut s = "pub(".to_owned();
                     if path.data.len() != 1
                         || (first_name != kw::SelfLower && first_name != kw::Super)
                     {
-                        f.write_str("in ")?;
+                        s.push_str("in ");
                     }
-                    // modified from `resolved_path()` to work with `DefPathData`
-                    let last_name = path.data.last().unwrap().data.get_opt_name().unwrap();
                     for seg in &path.data[..path.data.len() - 1] {
-                        write!(f, "{}::", seg.data.get_opt_name().unwrap())?;
+                        s.push_str(&format!("{}::", seg.data.get_opt_name().unwrap()));
                     }
-                    let path = anchor(vis_did, &last_name.as_str()).to_string();
-                    write!(f, "{}) ", path)
+                    s.push_str(&format!("{}) ", anchor));
+                    s
                 }
             }
-        })
+        };
+        display_fn(move |f| f.write_str(&to_print))
```

Comment:
Why did this change?

Reply:
Because of `TyCtxt<'tcx>`, and more precisely `'tcx`. This lifetime was a nightmare to go around...